service-broker/controller/controller.go

```diff
@@ -0,0 +1,179 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"fmt"
+	"log"
+	"sync"
+
+	"github.com/kubernetes-incubator/service-catalog/contrib/pkg/broker/controller"
+	"github.com/kubernetes-incubator/service-catalog/contrib/pkg/brokerapi"
+	"github.com/opensds/opensds/client"
+	"github.com/opensds/opensds/pkg/model"
+)
+
+type errNoSuchInstance struct {
+	instanceID string
+}
+
+func (e errNoSuchInstance) Error() string {
+	return fmt.Sprintf("no such instance with ID %s", e.instanceID)
+}
+
+type openSDSServiceInstance struct {
+	Name       string
+	Credential *brokerapi.Credential
+}
+
+type openSDSController struct {
+	cli         *client.Client
+	rwMutex     sync.RWMutex
+	instanceMap map[string]*openSDSServiceInstance
+}
+
+// CreateController creates an instance of an OpenSDS service broker controller.
+func CreateController(edp string) controller.Controller {
+	var instanceMap = make(map[string]*openSDSServiceInstance)
+	return &openSDSController{
+		cli: client.NewClient(&client.Config{
```

Comment:
It's better to refer OpenSDS Client like this.
https://github.com/opensds/nbp/blob/master/csi/server/plugin/opensds/controller.go#L26

Reply:
thx, I will update it later

Reply:
One question is all npb plugin could be containerized, so in this case how can you generate the environment variable? Could we also expose the ```endpoint``` variable to stdin through ```flag``` package?

Reply:
good point